src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/FrozenStringNode.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+using ILCompiler.DependencyAnalysisFramework;
+using Internal.TypeSystem;
+
+namespace ILCompiler.DependencyAnalysis
+{
+    public class FrozenStringNode : EmbeddedObjectNode, ISymbolNode
+    {
+        string _data;
+        int _syncBlockSize;
+
+        public FrozenStringNode(string data, TargetDetails target)
+        {
+            _data = data;
+            _syncBlockSize = target.PointerSize;
+        }
+
+        public string MangledName
+        {
+            get
+            {
+                return NodeFactory.CompilationUnitPrefix + "__Str_" + NodeFactory.NameMangler.GetMangledStringName(_data);
+            }
+        }
+
+        public override bool StaticDependenciesAreComputed
+        {
+            get
+            {
+                return true;
+            }
+        }
+
+        public override int Offset
+        {
+            get
+            {
+                // The frozen string symbol points at the EEType portion of the object, skipping over the sync block
+                return base.Offset + _syncBlockSize;
+            }
+        }
+
+        public override void EncodeData(ref ObjectDataBuilder dataBuilder, NodeFactory factory, bool relocsOnly)
+        {
+            dataBuilder.EmitZeroPointer(); // Sync block
+
+            //
+            // The GC requires a direct reference to frozen objects' EETypes. If System.String will be compiled into a separate
+            // binary, it must be cloned into this one.
+            //
+            if (factory.CompilationModuleGroup.ShouldReferenceThroughImportTable(factory.SystemStringClass))
+            {
+                dataBuilder.EmitPointerReloc(factory.ConstructedClonedTypeSymbol(factory.SystemStringClass)); // System.String EEType pointer
+            }
+            else
+            {
+                dataBuilder.EmitPointerReloc(factory.ConstructedTypeSymbol(factory.SystemStringClass)); // System.String EEType pointer
+            }
+
+            dataBuilder.EmitInt(_data.Length);
+
+            foreach (byte c in _data)
```

Comment:
Casting `char` to `byte` is only okay if all of our users are English speakers :)